Add readline environment variable support for bind -x commands

KingMob opened the pull request:
This adds support for setting and checking READLINE_LINE and READLINE_POINT env vars for arbitrary shell commands in bind -x.

The shell code can read them to learn what's currently on the line and where the cursor is. If either variable is changed, readline will update the command line/cursor.

This is compatible with bash 4, but not (fully) bash 5. Newer readline offers READLINE_ARGUMENT and READLINE_MARK as well.

----

Since this is the final piece of bind support (in Python/C anyway), I also cleaned up a lot of comments and debug printing.

andychu commented:
OK awesome, looks like it's passing all the tests now!

https://op.oilshell.org/uuu/github-jobs/9572/interactive.wwz/_tmp/spec/stateful/index.html

I will take a look at the code in more detail

I wouldn't be surprised if say running Warp finds some issues, or Koichi  finds some issues (since `ble.sh` is very thorough)

But this is great progress!

And we can also talk about C++ ... I think if we have some part of  Warp or ble.sh working, that will  be motivating

KingMob commented:
So, in an irony of ironies for my testing, it seems Warp is interfering with `bind -x` when a shell is "Warpified". 

I expected some of this, because part of how Warp offers nice GUI popups is by overriding standard keyseqs like Ctrl-r. And I'm not surprised that READLINE_LINE and READLINE_POINT `bind -x` commands fail, because I doubt they're using readline under the hood when they talk to the underlying bash process, and it looks like they haven't implemented support for a relatively-niche bash-specific feature.

What I *didn't* expect was that *all* bind -x bindings are blocked in Warpified mode, even custom ones that they shouldn't care about. I'm going to file it as a bug.

<img width="1769" alt="Screenshot 2025-05-19 at 12 41 41 PM" src="https://github.com/user-attachments/assets/81406280-a236-48c1-be25-d341742eb001" />

KingMob commented:
Also, I know Warp still uses a couple flags not yet supported in osh: `history -a` and `shopt -s histappend`

andychu commented on pyext/line_input.c:
```diff
@@ -1154,32 +1066,21 @@ on_bind_shell_command_hook(int count /* unused */, int key /* unused */) {
-    //     result = 1;
-    //     goto cleanup;
-    // }
+    update_line_if_needed(post_line_buffer);
```
What do these functions do?  Some comments would be useful

andychu commented on builtin/readline_osh.py:
```diff
@@ -49,46 +48,67 @@ def __exit__(self, type, value, traceback):
+        """Retrieve the value of an env var, return None if undefined"""
+
+        envvar_val = self.mem.GetValue(envvar_name, scope_e.GlobalOnly)
+        if envvar_val.tag() == value_e.Str:
```
To test if this value is either `Str` or `Undef` or else, please use this `with tagswitch` idiom

https://oils.pub/release/latest/doc/oils-repo/mycpp/README.html#with-tagstr_switch-switch-statement

You can also grep the code for `tagswitch` -- we use it all over

It translates to a switch statement in C++!

andychu commented on builtin/readline_osh.py:
```diff
@@ -49,46 +48,67 @@ def __exit__(self, type, value, traceback):
+        post_point = int(
+            readline_point) if readline_point is not None else point
+
+        self.mem.Unset(location.LName('READLINE_LINE'), scope_e.GlobalOnly)
```
I think we should use a context manager for unsetting to be "exception safe".    e.g. `eval.Run` may  raise an exception, and then this line won't be run

We have an idiom for this - `with ctx_Foo(...)`

https://oils.pub/release/latest/doc/oils-repo/mycpp/README.html#python-context-manager-c-constructor-and-destructor-raii

You can see many examples in say `core/state.py` -- the `__init__` pushes the state, and the `__exit__` pops it.  (And for our purposes, the `__enter__` must be there, but is ignored)

KingMob replied:
I'm adding it as `ctx_EnvVars`. It's in `readline_osh.py`, but it's general enough that it could be used elsewhere, if there's other places osh needs to temporarily set/unset env vars.

andychu commented on builtin/readline_osh.py:
```diff
@@ -49,46 +48,67 @@ def __exit__(self, type, value, traceback):
+            return None
+        else:
+            # bash has silent weird failures if you set the readline env vars
+            # to something besides a string, but I think an exception is better
```
Yeah this is a common issue -- bash doesn't check errors very well

Oh this is a good reason to be exception safe, which I commented about ...

Is the `TypeErr` caught somewhere?  Does it end up printing a message to `stderr`?

---

Sometimes GNU readline callbacks swallow errors, which is not great

KingMob replied:
As near as I can tell, the equivalent bash callback, `bash_execute_unix_command`, doesn't return a meaningful status code, so I doubt we should, either.

`bash_execute_unix_command` _does_ return 1 if it can't locate a keymap (rare, I suspect), but otherwise, it always returns 0, ignoring the return code of the bound shell commands. (The only thing the command return code is used for is to change the redrawing behavior when the return code is 124...whatever that's about.)

Just to make sure, I checked with bash 4 and 5. Running `bind -x` commands has no effect on `$?`, so it's not user-visible afaict. (Maybe ble.sh has more insight into this, though.)

----

Given this, I think there's no point in raising an exception, so I'm going to keep printing out the error, but remove the `raise`.

andychu replied:
OK that makes sense, generally I think printing more errors than bash makes sense, but failing hard sometimes leads to incompatibilities

andychu reviewed (commented):
Thank you, looks great!  Mostly style comments ...

KingMob commented:
OK, I addressed the feedback. See my comments in the 2 unresolved conversations.

----

Also, I fixed one minor missing piece that I just noticed at the very end. When deciding how to redraw after a bind -x command, we now check to ensure that the command return code isn't 124, to mimic bash.

andychu commented:
> Also, I know Warp still uses a couple flags not yet supported in osh: history -a and shopt -s histappend

I believe both of these have come up in the past, so are in scope ... thanks for mentioning it
